README.md
# stream-demo
Demonstrating connecting to the SpotGamma stream.

**NOTICE**: This API is not intended for SpotGamma subscribers.  If you are a SpotGamma subscriber who has stumbled onto this repository, this is intended for 3rd-party partners.

## REST APIs:
* [history](docs/history.md)
* [manifest](docs/manifest.md)

## Streams
Currently we support two separate streams of HIRO events: **Filtered** and **Absolute**.

**Filtered**: Runs a proprietary filter on options flow  and forms an opinion what "side" the transaction was with respect to Market Makers or Liquidity Providers.  This is  reflected in the sign of the delta value, where positive HIRO signal represents buying pressure and negative signal represents selling pressure.

**Absolute**: Filters no options and processes all options flow.  An absolute value is applied to all options greeks to reflect that we are forming no opinion on the "side" in relation to a market maker.

## Connecting
To connect, you'll first need to acquire a token by authenticating with `https://stream.spotgamma.com/auth` using a SpotGamma-provided username and password.

The stream is hosted over a websocket: `wss://stream.spotgamma.com/stream`.
Pass your authentication token in here as a GET parameter:
```js
const url = `wss://stream.spotgamma.com/stream?token=${token}`
```

## Subscribing
Once successfully connected, you can subscribe to individual underlying instruments or to all instruments using `*`.

**stream types**:
You can specify the streams to which you'd like to subscribe with the `stream_types` bitmask.

Possible values: `1` (filtered), `2` (absolute), or `3` (both streams).

**Filtered stream**
```js
ws.send(JSON.stringify({
  action: "subscribe",
  underlyings: ["*"], // listen to all underlyings
  stream_types: 1,    // only listen to the filtered stream
});
```

**Both streams**
```js
ws.send(JSON.stringify({
  action: "subscribe",
  underlyings: ["TSLA", "SPY", "AAPL"], // listen to selected underlyings
  stream_types: 3,                      // listen to both the filtered and "absolute" stream
});
```

## Unsubscribing
Send either an `"unsubscribe"` `action` and specify what you'd like to unsubscribe from or an `"unsusbscribe_all"` `action` for everything:
```js
ws.send(JSON.stringify({
  action: "unsubscribe",
  underlyings: ["TSLA"], // unsubscribe from `TSLA`
  stream_types: 3,       // unsubscribe from both streams
});
```

```js
ws.send(JSON.stringify({
  action: "unsubscribe_all", // unsubscribe everything: All symbols and all streams
});
```

## Messages
Messages come down as [msgpack'd](https://msgpack.org/index.html) tuples of the form:
```JavaScript
[streamType, signal]
```
where `streamType` can be either `1` for the **Filtered** stream or `2` for the **Absolute** stream.

`signal`, here, is itself a tuple (with an arity of 11) of the following format:
```js
[
  underlying, //  0: string
  timestamp,  //  1: 64-bit integer (UTC milliseconds)
  delta,      //  2: double - Delta HIRO value
  gamma,      //  3: double - Gamma HIRO value
  vega,       //  4: double - Vega HIRO value
  stockPrice, //  5: double - underlying stock price at the time the transaction took place
  sequence,   //  6: 64-bit integer
  expiry,     //  7: 64-bit integer (UTC milliseconds)
  strike,     //  8: float - strike price of the option
  flags,      //  9: unsigned 32-bit integer
  instrument, // 10: string - original option instrument namee
]
```

NOTE: You can rely on the combination of the `(underlying, sequenceID, strike, flags)` fields representing a unique HIRO event.

`flags` is a 32-bit bitmask containing auxiliary data in the following format:
```rust
const PUT_MASK: u32      = 0b000000000001; // 0x001 - 0 = call, 1 = put
const TYPE_MASK: u32     = 0b000000000110; // 0x006 - 0 = new, 1 = correction, 2 = cancel
const SIDE_MASK: u32     = 0b000000011000; // 0x018 - 0 = undefined, 1 = buy, 2 = sell
const NEXT_EXP_MASK: u32 = 0b000000100000; // 0x020 - 1 = expires in the next or nearest expiration
const RETAIL_MASK: u32   = 0b000001000000; // 0x040 - 1 = is a retail Time & Sale
const BLOCK_MASK: u32    = 0b000010000000; // 0x080 - 1 = is a block trade
const SPREAD_MASK: u32   = 0b000100000000; // 0x100 - 1 = is part of a spread leg
```


docs/history.md
# History REST API
The `/history` REST API provides up to a 2-day lookup history of all streamed events in both the `filtered` and `absolute` streams.
It takes 1 requried parameters and 3 optional query parameters:
1. `sym`: `String` (required)
2. `stream`: `String` (enum): values `absolute` or `filtered`. Defaults to `filtered`
3. `start`: `i64` UTC epoch in milliseconds (**inclusive**). Defaults to midnight UTC (`00:00+0:00`) today.
    * If it is a weekend day (Saturday or Sunday), defaults to the last Friday midnight UTC.
4. `end`: `i64` UTC epoch milliseconds (**exclusive**). Defaults to now

Example API call:

```sh
curl -H 'Authorization: Bearer <REDACTED>'  \
  'https://stream.spotgamma.com/history?sym=SPX&start=1693886400000&end=1693972799999&stream=absolute' -o- > 
  /tmp/spx.mp
```

Output is the same tuple structure as in the `stream` websocket API, but in one flat array or vector encoded in Messagepack format.

For example:
```sh
msgpack2json < /tmp/spx.mp | jq . | head -40
[
  [
    "SPX",
    1696392009047,
    -23696.807284663595,
    -1231.8660886464766,
    -106701.97455140528,
    4214.108910891089,
    "7285948200047870003",
    1696449600000,
    4290,
    48,
    ".SPXW231004C4290"
  ],
  [
    "SPX",
    1696392022690,
    -223024.91058721516,
    -3957.219741498195,
    -374487.26834143494,
    4214.1,
    "7285948258579382305",
    1696449600000,
    4225,
    48,
    ".SPXW231004C4225"
  ],
  [
    "SPX",
    1696392022691,
    -223024.91058721516,
    -3957.219741498195,
    -374487.26834143494,
    4214.1,
    "7285948258583576611",
    1696449600000,
    4225,
    48,
    ".SPXW231004C4225"
  ],
```



javascript/index.js
const { decode } = require("@msgpack/msgpack");

// BROWSER NOTE: Remove the definition below when running in a browser
const WebSocket = require("ws"); 

const SYMBOLS = ["__FAKE__", "/ESU23:XCME", "SPY", "AAPL", "TSLA"];

const Index = {
  UNDERLYING: 0,
  TIMESTAMP: 1,
  DELTA: 2,
  GAMMA: 3,
  VEGA: 4,
  STOCK_PRICE: 5,
  SEQUENCE_ID: 6,
  EXPIRY: 7,
  STRIKE: 8,
  FLAGS: 9,
  INSTRUMENT: 10,
};

const Mask = {
  OPTION_TYPE: 0x1,
  TNS_TYPE: 0x6,
  SIDE_TYPE: 0x18,
  IS_NEXT_EXP: 0x20,
  IS_RETAIL: 0x40,
  IS_BLOCK: 0x80,
  IS_SPREAD: 0x100,
};

const TYPE_STRS = [
  "n", // new
  "r", // correction
  "c", // cancel
  "", // ERROR: This is an unexpected value
];

const typeToString = (tns_type) => TYPE_STRS[tns_type];

SIDE_STRS = [
  "u", // undefined
  "b", // buy
  "s", // sell
  "", // ERROR: This is an unexpected value
];

const sideToString = (side_idx) => SIDE_STRS[side_idx];

const parseFlags = (flags) => [
  flags & Mask.OPTION_TYPE ? "P" : "C",
  typeToString((flags & Mask.TNS_TYPE) >> 1),
  sideToString((flags & Mask.SIDE_TYPE) >> 3),
  !!(flags & Mask.IS_NEXT_EXP) ? '1' : '0',
  !!(flags & Mask.IS_RETAIL) ? '1' : '0',
  !!(flags & Mask.IS_BLOCK) ? '1' : '0',
  !!(flags & Mask.IS_SPREAD) ? '1' : '0',
];

async function authenticateAndConnect() {
  let token = process.env.TOKEN;
  if (token == null) {
    const user = process.env.USER;
    const pass = process.env.PASSWORD;
    const encodedUserPass = btoa(`${user}:${pass}`);
    const authResult = await fetch("https://stream.spotgamma.com/auth", {
      headers: {
        Authorization: `Basic ${encodedUserPass}`,
      },
    });
    const jwt = await authResult.text();
    token = encodeURIComponent(jwt);
  }
  const socket = new WebSocket(
    `wss://stream.spotgamma.com/stream?token=${token}`
  );
  socket.addEventListener("open", function (_evt) {
    console.log("Websocket connected. Subscribing...");
    const msg = {
      action: "subscribe",
      underlyings: SYMBOLS,
      stream_types: 1, // Listen to filtered streeam
    };
    socket.send(JSON.stringify(msg));
  });

  let numMsgs = 0;
  socket.addEventListener("message", async function (event) {
    const { data } = event;
    if (typeof data === 'string') {
      console.log(typeof data);
      console.warn("Ack from server: ", data);
    } else {
      numMsgs += 1;
      // BROWSER NOTE: In a browser, we'd use the blob.arrayBuffer() function:
      //   i.e. const signalTuple = decode(await data.arrayBuffer());
      const signalTuple = decode(data, {
        useBigInt64: true, // for decoding 64-bit sequence ID's
      });

      const [stream, signal] = signalTuple;
      const flagFields = parseFlags(signal[Index.FLAGS]);
      const csvRow = signal
        .slice(Index.UNDERLYING, Index.SEQUENCE_ID)
        .concat(flagFields, [signal[Index.INSTRUMENT]])
        .join(",");
      console.log(csvRow);
    }
  });

  setInterval(() => {
    console.warn("Messages received: ", numMsgs);
  }, 10_000);

  socket.addEventListener("error", function (event) {
    console.error("error event:", event);
  });

  socket.addEventListener("close", function (event) {
    console.warn("Socket closed:", event);
  });
}

(async () => {
  await authenticateAndConnect();
})();


python/main.py
import base64
import json
import msgpack
import sys
import os
from urllib.request import Request, urlopen
from websockets.sync.client import connect

# Stream types
FILTERED = 1  # Sends filtered results where a side is determined
ABSOLUTE = 2  # Sends *all* TimeAndSales with all positive greeks.

OPTION_TYPE_MASK = 0x1
TNS_TYPE_MASK = 0x6
SIDE_TYPE_MASK = 0x18
IS_NEXT_EXP_MASK = 0x20
IS_RETAIL_MASK = 0x40
IS_BLOCK_MASK = 0x80
IS_SPREAD_MASK = 0x100

# Provide the USER and PASSWORD in environment variables
user = (os.getenv("USER") or "").encode("ascii")
password = (os.getenv("PASSWORD") or "").encode("ascii")

# Authenticate with https://stream.spotgamma.com/auth
req = Request("https://stream.spotgamma.com/auth")
ascii_auth = base64.b64encode(b"%s:%s" % (user, password)).decode("ascii")
req.add_header("Authorization", f"Basic {ascii_auth}")
token = urlopen(req).read().decode("ascii")
print(f"Auth token: {token}", file=sys.stderr)

subscription = json.dumps(
    {
        "action": "subscribe",
        "stream_types": FILTERED,  # Filtered only
        "underlyings": ["__FAKE__", "/ESU23:XCME", "SPY", "AAPL", "TSLA"],
    }
)

TYPE_STRS = [
    "n",  # new
    "r",  # correction
    "c",  # cancel
    "",  # ERROR: This is an unexpected value
]


def type_to_string(tns_type):
    return TYPE_STRS[tns_type]


SIDE_STRS = [
    "u",  # undefined
    "b",  # buy
    "s",  # sell
    "",  # ERROR: This is an unexpected value
]


def side_to_string(side_idx):
    return SIDE_STRS[side_idx]


def parse_flags(flags):
    option_type = "P" if (flags & OPTION_TYPE_MASK) else "C"
    tns_type = type_to_string((flags & TNS_TYPE_MASK) >> 1)
    side_type = side_to_string((flags & SIDE_TYPE_MASK) >> 3)
    is_next_exp = (flags & IS_NEXT_EXP_MASK) > 0
    is_retail = (flags & IS_RETAIL_MASK) > 0
    is_block = (flags & IS_BLOCK_MASK) > 0
    is_spread = (flags & IS_SPREAD_MASK) > 0
    return (
        option_type,
        tns_type,
        side_type,
        is_next_exp,
        is_retail,
        is_block,
        is_spread,
    )


print(subscription, file=sys.stderr)
with connect(f"wss://stream.spotgamma.com/stream?token={token}") as ws:
    ws.send(subscription)
    msg = ws.recv()
    while True:
        if type(msg) is bytes:
            (stream_type, signal) = msgpack.unpackb(msg, use_list=False)
            (
                underlying,
                timestamp,
                delta,
                gamma,
                vega,
                price,
                tns_id,
                _expiry,
                _strike,
                flags,
                instrument,
            ) = signal
            (
                option_type,
                tns_type,
                side,
                is_next_exp,
                is_retail,
                is_block,
                is_spread,
            ) = parse_flags(flags)
            if stream_type == FILTERED:
                print(
                    "%s,%s,%d,%f,%f,%f,%f,%d,%s,%s,%s,%d,%d,%d,%d"
                    % (
                        underlying,
                        instrument,
                        timestamp,
                        delta,
                        gamma,
                        vega,
                        price,
                        tns_id,
                        option_type,
                        tns_type,
                        side,
                        is_next_exp,
                        is_retail,
                        is_block,
                        is_spread,
                    )
                )
        else:
            print(msg, file=sys.stderr)
        msg = ws.recv()
